content/post/simulating-event-windows.md

```diff
@@ -0,0 +1,171 @@
+
+We set up our Wallaroo topology as follows
+
+```
+def application_setup(args):
+    in_host, in_port = wallaroo.tcp_parse_input_addrs(args)[0]
+    out_host, out_port = wallaroo.tcp_parse_output_addrs(args)[0]
+
+    ab = wallaroo.ApplicationBuilder("Apache Log file analysis")
+    ab.new_pipeline("convert",
+                    wallaroo.TCPSourceConfig(in_host, in_port, Decoder()))
+    ab.to(Count)
+    ab.to_sink(wallaroo.TCPSinkConfig(out_host, out_port, Encoder()))
+    return ab.build()
+```
+
+#### Decoder
+
+Our decoder needs to be able to correctly identify the `END_OF_DAY` message, and translate that into an appropriate message that will signal to the computations that they need to roll over to the next day. We will call this a `BoundaryMessage`, and all other messages will be `LogLine`s.
+
+```
+class Decoder(object):
+    def header_length(self):
+        return 4
+
+    def payload_length(self, bs):
+        return struct.unpack(">I", bs)[0]
+
+    def decode(self, bs):
+        if bs == 'END_OF_DAY':
+            return BoundaryMessage()
+        else:
+            return LogLine(bs)
+```
+
+#### Computation
+
+The computation itself needs to be able to do the following things:
+
+ - On the first `LogLine` for each window, determine which day we are counting for
+ - On all `LogLine`s, accumulate the counts
+ - On receipt of a `BoundaryMessage`, send all the output required as a
+   `SummaryMessage`
+
+```
+class Count(object):
+
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
+        self.current_batch = {}
+        self.current_day = None
+
+    def compute(self, data):
+        if type(data) is BoundaryMessage:
+            r = self.current_batch
+            self.reset()
+            return r
+        elif type(data) is LogLine:
+            return_code = determine_return_code(data)
+            self.current_batch[return_code] = self.current_batch.get(return_code, 0) + 1
+            if self.current_day is None:
+                self.current_day = determine_day(data)
+        else:
+            raise
+
+    def name(self):
+        return "Count return codes"
+
+    def process_batch(self, batch_data):
+        return len(batch_data)
+```
+
+#### Encoder
+
+On receipt of a `SummaryMessage`, we need to unpack the data inside it and
+create the lines of output.
+
+```
+class Encoder(object):
+    def encode(self, data):
+        return struct.pack('>II', 4, data)
+```
+
+## Next steps
+
+There are obvious limitations to this basic example. For instance, there is no partitioning. Additionally, we did not use the builtin stateful mechanisms in Wallaroo. All of this functionality can be added to production-level code, but for the purpose of illustrating how to simulate windows, we preferred to narrow the focus and reduce distractions.
```

Comment:
I feel like the skipping of using Wallaroo state is more harm than its worth. If someone copy/pastes this code and starts editing it into their own application, then they'd run into trouble with Wallaroo's state support not working on their (non-wallaroo) state.
Is it really that much more code to use wallaroo's state objects?